### PR TITLE
Switch our TypeScript build to top-level using references

### PR DESCRIPTION
**Pull request by rgbkrk**
Start building with `tsc` at the top level.

Attempting to follow the lead in https://github.com/RyanCavanaugh/learn-a for building our packages.

To build with this setup, use:

```
npx tsc -b packages
```

Make sure to run `yarn` first when you clone this branch.

## Goal

Get package builds and package build watches to be able to be run as one singular `tsc` command under the hood:

```
tsc -b packages --watch
```

### CODE REVIEW

**Comment by rgbkrk** [comment]
```
error TS6307: File '/home/circleci/nteract/packages/messaging/index.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/messaging/src/index.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/messaging/src/messages.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/messaging/src/types.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/rx-jupyter/index.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/rx-jupyter/src/base.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/rx-jupyter/src/contents.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/rx-jupyter/src/index.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/rx-jupyter/src/kernels.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/rx-jupyter/src/kernelspecs.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/rx-jupyter/src/sessions.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/rx-jupyter/src/terminals.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/commutable/src/cells.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/commutable/src/index.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/commutable/src/notebook.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/commutable/src/outputs.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/commutable/src/primitives.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/commutable/src/structures.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/commutable/src/v3.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/commutable/src/v4.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/actions/index.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/actions/index.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/epics/index.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

error TS6307: File '/home/circleci/nteract/packages/reducers/index.ts' is not in project file list. Projects must list all files or use an 'include' pattern.

```

**Comment by rgbkrk** [comment]
`@RyanCavanaugh` - ~~do you know what else we might be missing for our tsconfigs to make this work well?~~

EDIT: Nevermind, we're good now. Was hard to track down errors / figure out which packages were missing dependencies and references.

**Comment by willingc** [comment]
@rgbkrk Have you tried putting `tsc.js --b --v --f --listFiles` in circleci config for a bit more info?

**Comment by rgbkrk** [comment]
Your comment has unblocked me for figuring things out @willingc, thank you. I'm finding and fixing each of the bits we missed during the project reference setup. Thank you!

**Comment by rgbkrk** [comment]
Circle CI has been running this top level `tsc -b packages` command for 6 minutes and counting now. I am afraid.

**Comment by rgbkrk** [comment]
Oh ha, it finished at 6:08.

This will pass!

**Comment by rgbkrk** [comment]
It passes! Ready for independent review. @lgeiger 🙇

**Comment by rgbkrk** [comment]
In the meantime, I'll do a self-review.

* [x] `lib` directories for each package look right.
* [x] dependencies all installed

File: package.json [hunk 1]
```diff
@@ -30,13 +30,13 @@
...
-    "build:clean": "lerna run build:clean --parallel",
+    "build:clean": "npm run build:packages -- --clean",
     "build:apps": "lerna run build --scope nteract --scope @nteract/play --scope @nteract/commuter --scope nteract-on-jupyter --parallel --stream",
     "build:desktop": "lerna run build --scope nteract --stream",
     "build:desktop:watch": "lerna run build:watch --scope nteract --stream",
-    "build:packages": "NODE_ENV=production lerna run build --ignore @nteract/play --ignore @nteract/commuter --ignore nteract --ignore nteract-on-jupyter",
-    "build:packages:ci": "NODE_ENV=production lerna run build --ignore @nteract/play --ignore @nteract/commuter --ignore nteract --ignore nteract-on-jupyter",
-    "build:packages:watch": "lerna run build:lib:watch --parallel --ignore @nteract/play --ignore @nteract/commuter --ignore nteract --ignore nteract-on-jupyter",
+    "build:packages": "tsc -b packages",
+    "build:packages:ci": "npm run build:packages",
```
Comment (rgbkrk) [inline package.json]: We should make this `npm run build:packages -- --verbose` for our future selves.

File: packages/actions/index.ts [hunk 2]
```diff
@@ -1,4 +0,0 @@
-/**
- * @module actions
- */
-export * from "./src";
```
Comment (rgbkrk) [inline packages/actions/index.ts:4]: These `index.ts` files at the root of each package were not necessary (we resolve from `lib`), and weren't doing anything.

File: packages/transform-dataresource/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  "compilerOptions": {
+    "outDir": "lib",
+    "rootDir": "src"
+  },
+  "include": ["src/**/*.ts"],
+  "references": [
+      { "path": "../octicons" },
+      { "path": "../styled-blueprintjsx" }
+  ]
+}
```
Comment (rgbkrk) [inline packages/transform-dataresource/tsconfig.json]: I just noticed this one is probably in here erroneously and that we'll need to add it for the data explorer TypeScript migration PR.

File: packages/types/tsconfig.json [hunk 4]
```diff
@@ -1,7 +1,11 @@
 {
   "extends": "../../tsconfig.base.json",
   "compilerOptions": {
-    "outDir": "lib"
+    "outDir": "lib",
+    "rootDir": "src"
   },
-  "include": ["src"]
+  "include": ["src/**/*.ts"],
```
Comment (rgbkrk) [inline packages/types/tsconfig.json]: I'm noticing we can switch all the includes back to `src`. I'm going to go ahead and do that.

Comment (rgbkrk) [reply packages/types/tsconfig.json]: Done.

File: packages/transform-dataresource/tsconfig.json (hunk 3 above)
Comment (rgbkrk) [reply packages/transform-dataresource/tsconfig.json]: I'm going to delete it for now so there's not a merge conflict with the data explorer PR.

File: package.json (hunk 1 above)
Comment (rgbkrk) [reply package.json]: Done, will push shortly.

**Comment by rgbkrk** [comment]
🎉

**Comment by lgeiger** [comment]
@rgbkrk This looks great! A lot cleaner than building packages separately using lerna.

Do you think we should remove the build scripts from the individual [`packages/*/package.json`](https://github.com/nteract/nteract/blob/master/packages/mathjax/package.json#L11-L18)?

**Comment by jdetle** [comment]
@lgeiger, I think that is a good idea considering most can't really be built without other packages. But it can probably be in a follow up PR.

**Comment by rgbkrk** [comment]
Well... each of these packages can be built separately.

```
tsc -b packages/ansi-to-react
```

works as does:

```
yarn build:only @nteract/actions
```

**Comment by rgbkrk** [comment]
The project references don't work outside the monorepo, but I think that's ok.

**Comment by lgeiger** [comment]
> Well... each of these packages can be built separately.

Using `tsc -b packages/ansi-to-react` it work without any additional scripts in `packages/ansi-to-react/package.json` too.

**Comment by jdetle** [comment]
I totally think its okay too, I just wanted to make sure I understood how things were working.

**Pull request merged**